clients/service/src/lib.rs

```diff
@@ -169,10 +169,22 @@ impl<Config: Clone + Send + 'static, F: Fn()> ConnectionManager<Config, F> {
 	}
 }
 
-pub async fn wait_or_shutdown<F, E>(shutdown_tx: ShutdownSender, future2: F) -> Result<(), E>
+pub async fn wait_or_shutdown<F, E>(
+	shutdown_tx: ShutdownSender,
+	future2: F,
+	// a consumer that receives a precheck signal to start a task.
+	precheck_signal: Option<tokio::sync::broadcast::Receiver<()>>
+) -> Result<(), E>
 where
 	F: Future<Output = Result<(), E>>,
 {
+	if let Some(mut precheck_signal) = precheck_signal {
+		if let Err(e) = precheck_signal.recv().await {
```

Comment:
start the thread ONLY if precheck is finished.

Reply:
I forgot to reply to this.. 
I did not use a timeout, but the `try_recv()`; if it's not ready, then the loop will continue. 
The loop ends until it is  a success OR an error happens.